RPP: Add Psalm exception handling to src

RadoslavGeorgiev opened the pull request:
Fixes #6953

#### Changes proposed in this Pull Request

https://github.com/Automattic/woocommerce-payments/pull/6908 introduced exceptions within the new `src` directory, which triggered a discussion around exceptions and catching them.

@marcinbot suggested Psalm's `checkForThrowsDocblock` configuration, which turned out to use full analysis, and report uncaught exceptions from nested methods, not just one thrown in the current method. This PR implements it.

Details:

- Configuration is added to `psalm.xml` to check for exceptions.
- `includes` is excluded from the new check, as there are 287 issues to fix there (I was actually expecting more).
- Even though `includes` is excluded, it was preventing Psalm from working. Turns out that `WCPay\Exceptions\Order_Not_Found_Exception` was used in a few places without `use` statements for it. Those are fixed here.
- The `ExampleService` and `ExampleServiceWithDependencies` classes have exceptions and handlers now, to test the configuration.

#### Testing instructions

1. All checks here should be green.
2. Remove the `@throws` tag and the `try - catch` statement from `src/Internal/Service/ExampleServiceWithDependencies.php`.
3. Execute `npm run psalm` locally, and observe the errors.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (no tests really, this is not PHP)

htdat committed:
Add PaymentService and index files

htdat committed:
Exclude src from WP filename

htdat committed:
Add folder `src` to npm run release

htdat committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Adding the autoloading configuration

htdat committed:
Add changelog

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Removing test left-overs

htdat committed:
Change to PaymentProcessingService, use singular for Service

RadoslavGeorgiev committed:
First changes

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Update changelog/rpp-6677-psr-autoloader

Co-authored-by: Dat Hoang <htdat@users.noreply.github.com>

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge fix

RadoslavGeorgiev committed:
Switching to a unit test instead of a manual one

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Introducing the `lib` directory.

Because Mozart is a dev dependency, and not available within actions. This means that the lib directory will only be manipulated locally.

RadoslavGeorgiev committed:
Giving lib its own Composer instance

RadoslavGeorgiev committed:
Making psalm work.

The way we load files for Psalm prevents woocommerce-payments.php from defining functions, which would be otherwise available.

RadoslavGeorgiev committed:
Adding service providers and tests

RadoslavGeorgiev committed:
Tests part 1: Adding `wcpay_get_test_container()`.

RadoslavGeorgiev committed:
Allowing replacements during tests.

RadoslavGeorgiev committed:
Adding the container exception.

RadoslavGeorgiev committed:
Minor cleanups.

RadoslavGeorgiev committed:
Updating the lib readme

RadoslavGeorgiev committed:
Initial addition of the LegacyContainer, HooksProxy, and LegacyProxy.

RadoslavGeorgiev committed:
Fixing some tests.

RadoslavGeorgiev committed:
Switching to an example service for tests.

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
WIP: Replacements for delegates.

RadoslavGeorgiev committed:
Simplifying replacements

Instead of manipulating the complicated inner Container, ExtendedContainer now stores replacements locally.

RadoslavGeorgiev committed:
Fixing a unit test mistake

RadoslavGeorgiev committed:
Fixing Psalm issues

RadoslavGeorgiev committed:
Moving the Woo delegate container to a separate directory

RadoslavGeorgiev committed:
Testin improvement, incl. replacements.

Instead of overwriting existing definitions, ExtendedContainer adds a layer of replacements on top of the League container. This way references are kept after reset.

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Fixing lints

RadoslavGeorgiev committed:
Removing proxies, focusing the PR on the legacy container.

RadoslavGeorgiev committed:
Better tests:

- Adding a new `ExampleServiceWithDependencies` class, which tests all types of dependencies. `ExampleService` is just a placeholder now.
- Switched delegates with mocks for the containe tests. This way it's unit tests, rather than integration ones.
- Added a new test, which simulates all sorts of dependencies (`src, `includes`, and WooCommerce).

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' of github.com:Automattic/woocommerce-payments into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Switching namespaces

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Replacing WooPayments with WCPay

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Docs

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Switching to 4 spaces within composer.json

RadoslavGeorgiev committed:
Fixing an example

RadoslavGeorgiev committed:
Moving the initialization of the container within wcpay_get_container

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Use the container for the payment processing service

RadoslavGeorgiev committed:
Merge branch 'rpp/6678-dependency-container' into add/6707-container-additions

RadoslavGeorgiev committed:
Adding blank proxies, and a provider.

RadoslavGeorgiev committed:
Adding empty tests

RadoslavGeorgiev committed:
Working legacy proxy and tests

RadoslavGeorgiev committed:
Adding the hooks proxy

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Adding tests for the legacy container, and making it final.

RadoslavGeorgiev committed:
Adding docs

RadoslavGeorgiev committed:
Fixing all the broken tests

RadoslavGeorgiev committed:
Adding the available get_* instances from WC_Payments

RadoslavGeorgiev committed:
Merge branch 'add/6707-container-additions' into add/6707-proxies

RadoslavGeorgiev committed:
Adding docs

RadoslavGeorgiev committed:
Using has_global within get_global

RadoslavGeorgiev committed:
Updated pure methods wording

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-container-additions

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Merge branch 'add/6707-container-additions' into add/6707-proxies

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-container-additions

RadoslavGeorgiev committed:
Fixing typos and adding the correct return type.

RadoslavGeorgiev committed:
Throwing exceptions within get()

RadoslavGeorgiev committed:
Merge branch 'add/6707-container-additions' into add/6707-proxies

RadoslavGeorgiev committed:
Fixing a typo and some wording.

RadoslavGeorgiev committed:
Adding a ProxyException

RadoslavGeorgiev committed:
Running Psalm within src

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-proxies

RadoslavGeorgiev committed:
Initial commit.

- Adding some more examples to the example classes in order to test.
- Adding the checkForThrowsDockblock as configuration to Psalm.
- Disabling the new check for includes.
- MUST FIX: Disabling psalm for `includes`.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6954` or branch name `add/psalm-exceptions-to-src` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/psalm-exceptions-to-src) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: ba2a06fa23ed9b6b858ddb0a48b11671182133cb
- Build time: 2023-08-11 04:43:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev committed:
Fixing Psalm in includes, ignoring exceptions there.

- There was an exception, which was used in DocBlocks, without a `use` statement, which confused Psalm.
- Ignoring @throws exceptions for includes, that's nearly impossible to fix.

RadoslavGeorgiev committed:
Fixing the intentional src errors

RadoslavGeorgiev committed:
Merge branch 'develop' into add/6707-proxies

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Fixing a syntax error

RadoslavGeorgiev committed:
Merge branch 'add/6707-proxies' into add/psalm-exceptions-to-src

RadoslavGeorgiev committed:
Merge branch 'develop' into add/psalm-exceptions-to-src

htdat committed:
Merge branch 'develop' into add/psalm-exceptions-to-src

htdat reviewed (approved):
Pre-approve PR. Just a minor question but you are welcome to ignore it, and merge PR.

htdat commented on src/Internal/Service/ExampleServiceWithDependencies.php:
```diff
@@ -57,9 +58,17 @@ public function __construct(
 	 * @return bool
 	 */
 	public function is_in_test_mode() {
+		try {
+			$this->example_service->do_something_with_exception();
+		} catch ( Exception $e ) {
+			// Do nothing, this is fine. We're just testing Psalm.
+			return false;
+		}
+
```
I understand this is just for testing Psalm, but for which purpose? 

- With it, `return $this->mode->is_test();` is never reached. 
- And without it, Psalm catches up correctly that this method needs to declare `@throw`.

RadoslavGeorgiev replied:
It's just an example, which ensures that Psalm reads both `src` and `includes`, and understands where exceptions might be thrown. The purpose here isn't valid code, I just wanted to make sure that there are a couple of lines, which when commented out, would yield something like this:

<img width="864" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/5c67bb77-598a-46ad-8acc-f826fc471939">

Of course, I cannot commit without the `@throws` and `try-catch` because all checks would be 🔴 .

Anyway, those examples will be removed once we have better code to test, so I'm merging.
